### PR TITLE
Make clientApp and lang into string types and default to ''

### LINKED ISSUE #10121: Consider making clientApp and lang a string with a default value of ''
As discussed on Slack, there is an issue with our Flow types for `clientApp` and `lang`, which are stored in the `api` state. This has come to light with the patch for https://github.com/mozilla/addons-frontend/issues/10120. Here's a brief explanation:

In many components, we use `clientApp` and/or `lang`, the value for which we get from the `api` state. These are provided to the component via the `mapStateToProps` function, which gets these values from `state.api.clientApp` and `state.api.lang`, respectively. In the `api` state, these variables have a type of `string | null`, mainly because their initial state is `null`. However, anywhere in our code where we reference these variables we expect them to be strings. In fact, the way we have defined our Flow types for them say that they *will be* strings. So there is a disconnect between the types returned by `mapStateToProps` and the types that our component expects. This is something that has come to light as a result of generating/working on the patch for #10120.

We really should fix this issue, where the two instances of this variable (one from `mapStateToProps` and one that is defined in the `Props` for the component) have different types defined. It is incorrect. One problem with fixing this, though, is that it will have an impact on our code. Anywhere were we current define a prop of `clientApp` as `string` would need to be changed to `string | null`. That would often result in us having to add extra code into the component to deal with the fact that `clientApp` can actually be `null`. None of that code is currently present, as we just assume `clientApp` will be a string, and it hasn't caused any errors (because it always *is* a string).

My proposal to deal with this issue is to change the types of `clientApp` and `lang` in the `api` state so that they are defined as `string` and to make their initial value `''`. If we do that we can fix all the type-disconnect issues, but won't have to add any extra code anywhere.

One downside to this, as pointed out by @willdurand, is that we would no longer be alerted to a situation in which one of these variables has somehow not been initialized. I don't believe that is happening anywhere in our code, or we'd already be seeing runtime errors, but it is a theoretical possibility.

I did search through the code to see if we have any code currently that specifically looks for a value of `null` for either `clientApp` or `lang` and I didn't find any, so I am pretty confident that we can make this change without affecting the app or introducing any bugs.

I will put up a patch to make this change and we can discuss it further either there or here.

### PR DESCRIPTION
**Pull request by bobsilverberg**
Fixes #10121

### CODE REVIEW

**Comment by bobsilverberg** [comment]
Requesting review from @willdurand as we've already discussed this extensively, much of which is documented in the issue

File: src/amo/reducers/api.js [hunk 1]
```diff
@@ -101,9 +102,7 @@ export type SetClientAppAction = {|
 |};
 
 export function setClientApp(clientApp: string): SetClientAppAction {
-  if (!clientApp) {
-    throw new Error('clientApp cannot be falsey');
-  }
+  invariant(clientApp, 'clientApp cannot be falsey');
```
Comment (willdurand) [inline src/amo/reducers/api.js:105]: We probably want an `invariant` in `setLang` too.

**Review by willdurand — APPROVED** [review]
I think it's fine. If, for some reasons, we don't have a valid `clientApp` or `lang` in the UI layer (React components), we'll have bigger problems than just an empty value as it would mean something went wrong in Redux (Saga).

r+wc

File: src/amo/reducers/api.js (hunk 1 above)
Comment (bobsilverberg) [reply src/amo/reducers/api.js:105]: Yes. I needed to remove a test and some code that would never have been executed to allow this, but it makes sense. I did some research and it looks like the code never should have been added, as it assumes that `lang` can be falsey but it really cannot be.

**Pull request merged**